From ea71c92640a871da306d555b588f42c8e5a90743 Mon Sep 17 00:00:00 2001
From: srinidhira0 <srinidhir@vmware.com>
Date: Tue, 8 Feb 2022 11:10:27 +0000
Subject: [PATCH] crypto: AF_ALG - add ECC support

The patch externalizes the ECC crypto API to user space. ECC domain does not
require cipher handler (such as ECDH for KPP) as ECC routines can be called
directly and there is no ECC cipher algo.

The following operations are supported:

 * Key generation. Return values are private key, public key in form of X and
   Y coordinates.
 * Key verification. It takes X and Y coordinates of a public key as an input
   and performs validation.

Since we need only 2 operations from ECC: to verify key and to generate
key, simply sendmsg/recvmsg API for this domain will be sufficient.

Enabling CONFIG_CRYPTO_USER_API_ECC is not recommended for production, but
it will not harm if enabled, as it does not expose used by kernel keys to
user space.

Added support for LKCM 5.0:

 * In LKCM 4.0 (kernel v5.10.x), nist_p256 was the only fips supported curve.
 * In kernel v6.1.x, there are two fips supported curves - nist_p256 and nist_p384
 * Tweaked ecc_sendmsg and ecc_recvmsg to accomodate both these curves while performing
   keyVer and keyGen respectively.
 * Make allocation of cipher handle fail if a curve other than nist_p256 and nist_p384
   is sent from user-space.

Signed-off-by: Alexey Makhalov <amakhalov@vmware.com>
Signed-off-by: srinidhira0 <srinidhir@vmware.com>
Signed-off-by: Srish Srinivasan <ssrish@vmware.com>
Signed-off-by: Shivani Agarwal <shivani.agarwal@broadcom.com>
---
 crypto/Kconfig     |   8 ++
 crypto/Makefile    |   1 +
 crypto/algif_ecc.c | 277 +++++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 286 insertions(+)
 create mode 100644 crypto/algif_ecc.c

diff --git a/crypto/Kconfig b/crypto/Kconfig
index 573d692f1..31fab9344 100644
--- a/crypto/Kconfig
+++ b/crypto/Kconfig
@@ -1429,6 +1429,14 @@ config CRYPTO_USER_API_KPP
 	  primitives algorithms. This covers Diffie-Hellman and EC
 	  Diffie-Hellman.
 
+config CRYPTO_USER_API_ECC
+	tristate "User-space interface for ECC primitives algorithms"
+	depends on NET
+	select CRYPTO_USER_API
+	help
+	  This option enables the user-spaces interface for ECC
+	  primitives algorithms that are used in EC Diffie-Hellman.
+
 config CRYPTO_HASH_INFO
 	bool
 
diff --git a/crypto/Makefile b/crypto/Makefile
index 0a5cb6e22..f52797a08 100644
--- a/crypto/Makefile
+++ b/crypto/Makefile
@@ -202,6 +202,7 @@ ecrdsa_generic-y += ecrdsa_pub_key.asn1.o
 obj-$(CONFIG_CRYPTO_ECRDSA) += ecrdsa_generic.o
 obj-$(CONFIG_CRYPTO_USER_API_AKCIPHER) += algif_akcipher.o
 obj-$(CONFIG_CRYPTO_USER_API_KPP) += algif_kpp.o
+obj-$(CONFIG_CRYPTO_USER_API_ECC) += algif_ecc.o
 
 #
 # generic algorithms and the async_tx api
diff --git a/crypto/algif_ecc.c b/crypto/algif_ecc.c
new file mode 100644
index 000000000..865b479f5
--- /dev/null
+++ b/crypto/algif_ecc.c
@@ -0,0 +1,277 @@
+/*
+ * algif_ecc: User-space interface for ECC routines
+ *
+ * Copyright (C) 2022, VMware Inc.
+ * Author: Alexey Makhalov <amakhalov@vmware.com>
+ *
+ * This file provides the user-space API for ECC routins for keyGen and keyVer
+ * ACVP testing.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * The following concept of the memory management is used:
+ *
+ * The kernel maintains two SGLs, the TX SGL and the RX SGL. The TX SGL is
+ * filled by user space with the data submitted via sendpage/sendmsg. Filling
+ * up the TX SGL does not cause a crypto operation -- the data will only be
+ * tracked by the kernel. Upon receipt of one recvmsg call, the caller must
+ * provide a buffer which is tracked with the RX SGL.
+ *
+ * During the processing of the recvmsg operation, the cipher request is
+ * allocated and prepared. As part of the recvmsg operation, the processed
+ * TX buffers are extracted from the TX SGL into a separate SGL.
+ *
+ * After the completion of the crypto operation, the RX SGL and the cipher
+ * request is released. The extracted TX SGL parts are released together with
+ * the RX SGL release.
+ */
+
+#include <crypto/if_alg.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/list.h>
+#include <linux/mm.h>
+#include <linux/module.h>
+#include <linux/net.h>
+#include <net/sock.h>
+#include <crypto/internal/ecc.h>
+#include "ecc_curve_defs.h"
+
+/*
+ * Sendmsg to ecc socket is used only to verify public key curve point.
+ * There are 2 input iov segments, one for x coordinate and another for y.
+ * On success, it returns number of bytes that were verified.
+ * Negative return value means failure. -EINVAL value get returned if
+ * key verification failed.
+ */
+
+static int ecc_sendmsg(struct socket *sock, struct msghdr *msg, size_t size)
+{
+	const struct iovec *iov;
+	int ret;
+	/* FIPS supported curves: nist_p256 and nist_p384 */
+	const struct ecc_curve *curve = NULL;
+	u8 x_buf[48] = {0}, y_buf[48] = {0};
+	int curve_id = 0;
+
+	if (!iter_is_iovec(&msg->msg_iter))
+		return -ENOTSUPP;
+
+	if(msg->msg_iter.nr_segs == 3) {
+
+		iov = msg->msg_iter.iov;
+		struct ecc_point pk;
+
+		if(iov[2].iov_len == 256) {
+			curve = &nist_p256;
+			curve_id = 2;
+		} else if(iov[2].iov_len == 384) {
+			curve = &nist_p384;
+			curve_id = 3;
+		} else if(iov[2].iov_len == 192) {
+			curve = &nist_p192;
+			curve_id = 1;
+		} else
+			return -EINVAL;
+
+		if (iov[0].iov_len != curve->g.ndigits * sizeof(u64) ||
+	    		iov[1].iov_len != curve->g.ndigits * sizeof(u64))
+			return -EINVAL;
+
+		ret = copy_from_user(x_buf, iov[0].iov_base, iov[0].iov_len);
+		if (ret) {
+			pr_err("\n Error in copying X from iov_base %p %d\n",
+					iov[0].iov_base, ret);
+			return -ENOMEM;
+		}
+		ret = copy_from_user(y_buf, iov[1].iov_base, iov[1].iov_len);
+		if (ret) {
+			pr_err("\n Error in copying Y from iov_base %p %d\n",
+						iov[1].iov_base, ret);
+			return -ENOMEM;
+		}
+
+		pk.ndigits = curve->g.ndigits;
+		pk.x = kmalloc(pk.ndigits * sizeof(u64), GFP_KERNEL);
+		if (!pk.x)
+			return -ENOMEM;
+
+		pk.y = kmalloc(pk.ndigits * sizeof(u64), GFP_KERNEL);
+		if (!pk.y) {
+			kfree(pk.x);
+			return -ENOMEM;
+		}
+
+		ecc_swap_digits((u64 *)x_buf, pk.x, pk.ndigits);
+		ecc_swap_digits((u64 *)y_buf, pk.y, pk.ndigits);
+
+		ret = ecc_is_pubkey_valid_full(curve, &pk);
+
+
+		kfree(pk.x);
+		kfree(pk.y);
+		return ret;
+	}
+	else
+		return -EINVAL;
+}
+
+/*
+ * Recvmsg from ecc socket is used only to generate private/public key pair
+ * There are 3 output iov segments, one for private key and other two for x
+ * and y coordinates of the public key. Concatenation of bytes from these
+ * coordinates gives a public key.
+ * On success, it returns number of bytes that were provided.
+ * Negative return value means failure.
+ */
+static int ecc_recvmsg(struct socket *sock, struct msghdr *msg,
+			    size_t ignored, int flags)
+{
+        const struct iovec *iov;
+        u64 pubkey[12];
+        int ret;
+        /* FIPS supported curves: nist_p256 and nist_p384 */
+        struct ecc_curve *curve = NULL;
+        u8 d_buf[48] = {0};
+        int curve_id = 0;
+
+        if (!iter_is_iovec(&msg->msg_iter))
+                return -ENOTSUPP;
+
+        if (msg->msg_iter.nr_segs != 4)
+                return -EINVAL;
+
+        iov = msg->msg_iter.iov;
+
+	if(iov[3].iov_len == 256) {
+		curve = &nist_p256;
+		curve_id = 2;
+	} else if(iov[3].iov_len == 384) {
+		curve = &nist_p384;
+		curve_id = 3;
+	} else if(iov[3].iov_len == 192) {
+		curve = &nist_p192;
+		curve_id = 1;
+	} else
+		return -EINVAL;
+
+        if (iov[0].iov_len != curve->g.ndigits * sizeof(u64) ||
+            iov[1].iov_len != curve->g.ndigits * sizeof(u64) ||
+            iov[2].iov_len != curve->g.ndigits * sizeof(u64))
+                return -EINVAL;
+
+        ret = ecc_gen_privkey(curve_id, curve->g.ndigits, (u64 *)d_buf);
+        if (ret < 0)
+                return ret;
+
+        ret = ecc_make_pub_key(curve_id, curve->g.ndigits, (u64 *)d_buf, pubkey);
+        if (ret < 0)
+                return ret;
+
+        ret = copy_to_user(iov[0].iov_base, d_buf, iov[0].iov_len);
+        if (ret) {
+                pr_err("\n Error in copying from iov_base %p %d\n",
+                                        iov[0].iov_base, ret);
+                return -ENOMEM;
+        }
+        ret = copy_to_user(iov[1].iov_base, pubkey,
+                                        curve->g.ndigits * sizeof(u64));
+        if (ret) {
+                pr_err("\n Err in copying to user %d\n", ret);
+                return -ENOMEM;
+        }
+        ret = copy_to_user(iov[2].iov_base, &pubkey[curve->g.ndigits],
+                                curve->g.ndigits * sizeof(u64));
+        if (ret) {
+                pr_err("\n Err in copying to user %d\n", ret);
+                return -ENOMEM;
+        }
+
+        return msg->msg_iter.count;
+}
+
+static struct proto_ops algif_ecc_ops = {
+	.family		=	PF_ALG,
+
+	.connect	=	sock_no_connect,
+	.socketpair	=	sock_no_socketpair,
+	.getname	=	sock_no_getname,
+	.ioctl		=	sock_no_ioctl,
+	.listen		=	sock_no_listen,
+	.shutdown	=	sock_no_shutdown,
+	.getsockopt	=	sock_getsockopt,
+	.mmap		=	sock_no_mmap,
+	.bind		=	sock_no_bind,
+	.accept		=	sock_no_accept,
+	.setsockopt	=	sock_setsockopt,
+
+	.release	=	af_alg_release,
+	.sendmsg	=	ecc_sendmsg,
+	.sendpage	=	af_alg_sendpage,
+	.recvmsg	=	ecc_recvmsg,
+	.poll		=	af_alg_poll,
+};
+
+static void *ecc_bind(const char *name, u32 type, u32 mask)
+{
+	if(!strcmp(name, "ecdh-nist-p192")  ||
+	   !strcmp(name, "ecdh-nist-p256")  ||
+	   !strcmp(name, "ecdh-nist-p384")  ||
+	   !strcmp(name, "ecdsa-nist-p192") ||
+	   !strcmp(name, "ecdsa-nist-p256") ||
+	   !strcmp(name, "ecdsa-nist-p384")) {
+		return 0;
+	}
+	return ERR_PTR(-EINVAL);
+}
+
+static void ecc_release(void *private)
+{
+}
+
+static void ecc_sock_destruct(struct sock *sk)
+{
+}
+
+static int ecc_accept_parent(void *private, struct sock *sk)
+{
+	sk->sk_destruct = ecc_sock_destruct;
+	return 0;
+}
+
+static const struct af_alg_type algif_type_ecc = {
+	.bind		=	ecc_bind,
+	.release	=	ecc_release,
+	.setkey		=	NULL,
+	.setpubkey	=	NULL,
+	.dhparams	=	NULL,
+	.ecdhcurve	=	NULL,
+	.setauthsize	=	NULL,
+	.accept		=	ecc_accept_parent,
+	.accept_nokey	=	NULL,
+	.ops		=	&algif_ecc_ops,
+	.ops_nokey	=	NULL,
+	.name		=	"ecc",
+	.owner		=	THIS_MODULE
+};
+
+static int __init algif_ecc_init(void)
+{
+	return af_alg_register_type(&algif_type_ecc);
+}
+
+static void __exit algif_ecc_exit(void)
+{
+	int err = af_alg_unregister_type(&algif_type_ecc);
+
+	BUG_ON(err);
+}
+
+module_init(algif_ecc_init);
+module_exit(algif_ecc_exit);
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Alexey Makhalov <amakhalov@vmware.com>");
+MODULE_DESCRIPTION("ECC primitives kernel crypto API user space interface");
-- 
2.35.6
